[quant][graphmode] Different rule for handling `aten::cat`

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38552 [quant] Make MinMaxObserver and MovingAverageMinMaxObserver work with List[Tensor]
* #38436 [quant] Histogram Observer support List[Tensor]
* #38667 [quant][graphmode] Different rule for add/add_/mul/mul_
* **#38570 [quant][graphmode] Different rule for handling `aten::cat`**

Summary:
We changed the rule of quantizing `aten::cat`, previously `aten::cat` is considered to be
an op that should always be quantized, like `aten::conv2d`, but this is not ideal, a better
way is to quantize the output of `aten::cat` depending on whether the input is quantized, if it is
then we'll quantize the output, if not, then we will not quantize the output, since `aten::cat` works both on
quantized and non-quantized tensor.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21600160](https://our.internmc.facebook.com/intern/diff/D21600160)

jerryzh168 committed:
[quant][graphmode] Different rule for handling `aten::cat`

Summary:
We changed the rule of quantizing `aten::cat`, previously `aten::cat` is considered to be
an op that should always be quantized, like `aten::conv2d`, but this is not ideal, a better
way is to quantize the output of `aten::cat` depending on whether the input is quantized, if it is
then we'll quantize the output, if not, then we will not quantize the output, since `aten::cat` works both on
quantized and non-quantized tensor.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Different rule for handling `aten::cat`"


Summary:
We changed the rule of quantizing `aten::cat`, previously `aten::cat` is considered to be
an op that should always be quantized, like `aten::conv2d`, but this is not ideal, a better
way is to quantize the output of `aten::cat` depending on whether the input is quantized, if it is
then we'll quantize the output, if not, then we will not quantize the output, since `aten::cat` works both on
quantized and non-quantized tensor.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21600160](https://our.internmc.facebook.com/intern/diff/D21600160)

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1412,7 +1419,20 @@ def forward(self, x, y):
+                   .run(m.graph)
+
+        # non quantized cat
+        m = torch.jit.script(NonQuantizedCat()).eval()
```
In this case, what is the expected behavior? Would you quantize all the inputs to non-quantized cat  and quantize the output too? In general, dont you quantize the input to a model?

jerryzh168 replied:
not we don't quantize the inputs to cat in this case, the output of cat is only quantized when all the inputs are quantized.

jerryzh168 replied:
we will not quantize the input of cat in this case.

raghuramank100 commented on torch/csrc/jit/passes/quantization/insert_observers.cpp:
```diff
@@ -1025,7 +1032,13 @@ InsertObserversHelper::insertObserversFor(
               !isObserved(v, block_observed_values)) {
-            if (auto observer_opt = getObserverFor(v)) {
+            auto observer_opt = getObserverFor(v);
+            // If the node is one of the propagate quant node, e.g.
```
Isnt this logic also true for add or conv? Why special case concat?

jerryzh168 replied:
This is only true for cat, for conv we'll always quantize its input and output

jerryzh168 replied:
this is only true for cat, for conv we'll always quantize its input and output

raghuramank100 replied:
We need a consistent set of rules, consider the following two models:
``` python
import torch
import torch.nn as nn
class testM(nn.Module):
    def __init__(self):
        super().__init__()
        self.c = nn.Conv2d(3,5,1)
        self.d = nn.Conv2d(3,5,1)
    
    def forward(self, x):
# If there is a nn.Identity or shape, will the inputs be quantized?
        y = self.c(x)
        z = self.d(x)
        w = torch.cat((y, z))
        return w

# Second one:
class testM2(nn.Module):
    def __init__(self):
        super().__init__()
        self.c = nn.Conv2d(3,5,1)
        self.d = nn.Conv2d(3,5,1)
    
    def forward(self, x):
        w = torch.cat((x, x))
        y = self.c(w)
        z = self.d(w)
        return w

```
In the first case, the input will be quantized. In the second case the input will not be. 
What if in the first case I have an nn.Identity prior to the conv or a reshape? In that case do we quantize the inputs?

jerryzh168 replied:
we don't quantize identity or reshape, we also don't accept input that's already quantized outside of the model.

jerryzh168 replied:
In terms user interface, user will always provide a floating point Tensor, regardless of how the model is quantized

raghuramank100 replied:
Got it, the input is always in fp and in certain cases the input is quantized (conv) and in certain cases it is not (cat)

jerryzh168 committed:
Update on "[quant][graphmode] Different rule for handling `aten::cat`"


Summary:
We changed the rule of quantizing `aten::cat`, previously `aten::cat` is considered to be
an op that should always be quantized, like `aten::conv2d`, but this is not ideal, a better
way is to quantize the output of `aten::cat` depending on whether the input is quantized, if it is
then we'll quantize the output, if not, then we will not quantize the output, since `aten::cat` works both on
quantized and non-quantized tensor.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21600160](https://our.internmc.facebook.com/intern/diff/D21600160)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Different rule for handling `aten::cat`"


Summary:
We changed the rule of quantizing `aten::cat`, previously `aten::cat` is considered to be
an op that should always be quantized, like `aten::conv2d`, but this is not ideal, a better
way is to quantize the output of `aten::cat` depending on whether the input is quantized, if it is
then we'll quantize the output, if not, then we will not quantize the output, since `aten::cat` works both on
quantized and non-quantized tensor.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21600160](https://our.internmc.facebook.com/intern/diff/D21600160)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Different rule for handling `aten::cat`"


Summary:
We changed the rule of quantizing `aten::cat`, previously `aten::cat` is considered to be
an op that should always be quantized, like `aten::conv2d`, but this is not ideal, a better
way is to quantize the output of `aten::cat` depending on whether the input is quantized, if it is
then we'll quantize the output, if not, then we will not quantize the output, since `aten::cat` works both on
quantized and non-quantized tensor.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21600160](https://our.internmc.facebook.com/intern/diff/D21600160)

[ghstack-poisoned]
